enterprise/server/ociregistry/ociregistry.go

```diff
@@ -0,0 +1,305 @@
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+)
+
+const (
+	rangeHeaderBytesPrefix = "bytes="
+
+	registryInstanceName = "registry_artifacts"
+)
+
+var (
+	manifestReqRE = regexp.MustCompile("/v2/(.+?)/manifests/(.+)")
+	blobReqRE     = regexp.MustCompile("/v2/(.+?)/blobs/(.+)")
+	blobUploadRE  = regexp.MustCompile("/v2/(.+?)/blobs/uploads/")
+
+	enableRegistry = flag.Bool("ociregistry.enabled", false, "Whether to enable registry services")
+)
+
+type registry struct {
+	env environment.Env
+}
+
+func Register(env environment.Env) error {
+	if !*enableRegistry {
+		return nil
+	}
+
+	mux := env.GetInternalHTTPMux()
+	if mux == nil {
+		return status.FailedPreconditionErrorf("Registry requires internal HTTP mux")
+	}
+
+	r, err := New(env)
+	if err != nil {
+		return err
+	}
+
+	mux.Handle("/v2/", r)
+	return nil
+}
+
+func New(env environment.Env) (*registry, error) {
+	r := &registry{
+		env: env,
+	}
+	return r, nil
+}
+
+func (r *registry) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	r.handleRegistryRequest(w, req)
+}
+
+func (r *registry) handleRegistryRequest(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	ctx, err := prefix.AttachUserPrefixToContext(ctx, r.env)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("could not attach user prefix: %s", err), http.StatusInternalServerError)
+		return
+	}
+	authHeader := req.Header.Get("Authorization")
+	log.CtxInfof(ctx, "%s %q '%s'", req.Method, req.RequestURI, authHeader)
+	// Clients issue a GET /v2/ request to verify that this  is a registry
+	// endpoint.
+	if req.RequestURI == "/v2/" {
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+	forwardedRegistry := req.Header.Get("X-Forwarded-Host")
+	// Request for a manifest or image index.
+	if m := manifestReqRE.FindStringSubmatch(req.RequestURI); len(m) == 3 {
+		imageName := m[1]
+		refName := m[2]
+		if forwardedRegistry != "" {
+			imageName = forwardedRegistry + "/" + imageName
+		}
+		r.handleManifestRequest(ctx, w, req, imageName, refName)
+		return
+	}
+	// Uploading a blob
+	if m := blobUploadRE.FindStringSubmatch(req.RequestURI); len(m) == 2 {
+		w.WriteHeader(http.StatusNotImplemented)
+		return
+	}
+	// Request for a blob (full layer or layer chunk).
+	if m := blobReqRE.FindStringSubmatch(req.RequestURI); len(m) == 3 {
+		imageName := m[1]
+		refName := m[2]
+		if forwardedRegistry != "" {
+			imageName = forwardedRegistry + "/" + imageName
+		}
+		r.handleBlobRequest(ctx, w, req, imageName, refName)
+		return
+	}
+	http.NotFound(w, req)
+}
+
+func (r *registry) handleManifestRequest(ctx context.Context, w http.ResponseWriter, req *http.Request, imageName, refName string) {
```

Comment:
Bit of a nit, but since this is kind of greenfield code could we add some simple comments on the important top level methods in here like handleManifestRequest, handleBlobRequest, + any others you think would be worthwhile?

Reply:
Not a nit! Just left comments on the key pieces without going overboard. The key thing to understand is that we're implementing just enough of the distribution spec to be able to pull public images.